### PR TITLE
Migrate `cdylib-dylib-linkage` `run-make` test to rmake

### PR DESCRIPTION
**Pull request by Oneirical**
Part of #121876 and the associated [Google Summer of Code project](https://blog.rust-lang.org/2024/05/01/gsoc-2024-selected-projects.html).

~~Those sysroot tests are always fun. I'm getting local errors that don't make a lot of sense about my own sysroot not existing, so I am trying this in CI to see what happens.~~

~~EDIT: I am getting the same error here. The strange thing is, when I try to navigate to `/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/lib/rustlib/x86_64-unknown-linux-gnu/lib` on my personal computer, the directory does exist, but the error message is that the directory does not.~~

EDIT 2: The sysroot path just needed to be trimmed!

Please try:

// try-job: x86_64-msvc // passed previously
try-job: x86_64-mingw
try-job: x86_64-gnu-llvm-18
try-job: i686-msvc
try-job: aarch64-apple

### CODE REVIEW

**Comment by jieyouxu** [comment]
I'll take a look at this (eventually™), I'm looking at some other PRs that also need some holy water.

**Comment by jieyouxu** [comment]
uh... does the sysroot path have a newline?

```
the directory in path "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2
/lib/rustlib/x86_64-unknown-linux-gnu/lib" could not be read: Os { code: 2, kind: NotFound, message: "No such file or directory" }
```

File: tests/run-make/cdylib-dylib-linkage/rmake.rs [hunk 1]
```diff
@@ -0,0 +1,39 @@
...
+use run_make_support::{
+    bin_name, cc, has_extension, has_prefix, has_suffix, is_msvc, name_contains, path, run, rustc,
+    shallow_find_files, target,
+};
+
+fn main() {
+    rustc().arg("-Cprefer-dynamic").input("bar.rs").run();
+    rustc().input("foo.rs").run();
+    let sysroot = rustc().print("sysroot").run().stdout_utf8();
+    let target_sysroot = path(sysroot).join("lib/rustlib").join(target()).join("lib");
```
Comment (jieyouxu) [inline tests/run-make/cdylib-dylib-linkage/rmake.rs:19]: Problem: yes, `--print sysroot` produces a path

```
/path/to/sysroot\n
```

notice the `\n`, you need to trim that off before you join 😄

Comment (Oneirical) [reply tests/run-make/cdylib-dylib-linkage/rmake.rs]: Wow, thank you, I can hardly believe it was that easy. I'm pretty sure I ran into this exact same error a while back, and that you gave me the same advice.

**Comment by Oneirical** [comment]
The dynamic library std filter was incorrect - it was only looking for files that are "not an rlib, with std in the name somewhere", which could include rmeta files. This new implementation explicitly looks for dylibs with std in their filename.

Please try again. @rustbot review

File: src/tools/run-make-support/src/path_helpers.rs [hunk 2]
```diff
@@ -78,3 +79,8 @@ pub fn has_extension<P: AsRef<Path>>(path: P, extension: &str) -> bool {
 pub fn has_suffix<P: AsRef<Path>>(path: P, suffix: &str) -> bool {
     path.as_ref().file_name().is_some_and(|name| name.to_str().unwrap().ends_with(suffix))
 }
+
+/// Returns true if the filename at `path` contains `needle`.
+pub fn name_contains<P: AsRef<Path>>(path: P, needle: &str) -> bool {
+    path.as_ref().file_name().is_some_and(|name| name.to_str().unwrap().contains(needle))
+}
```
Comment (jieyouxu) [inline src/tools/run-make-support/src/path_helpers.rs:86]: Discussion: `name_contains` -> `filename_contains`?

File: tests/run-make/cdylib-dylib-linkage/rmake.rs [hunk 3]
```diff
@@ -0,0 +1,40 @@
...
+    rustc().input("foo.rs").run();
+    let sysroot = rustc().print("sysroot").run().stdout_utf8();
+    let sysroot = sysroot.trim();
+    let target_sysroot = path(sysroot).join("lib/rustlib").join(target()).join("lib");
+    if is_msvc() {
+        let mut libs = shallow_find_files(&target_sysroot, |path| {
+            has_prefix(path, "libstd-") && has_suffix(path, ".dll.lib")
+        });
+        libs.push(path("foo.dll.lib"));
+        libs.push(path("bar.dll.lib"));
```
Comment (jieyouxu) [inline tests/run-make/cdylib-dylib-linkage/rmake.rs:26]: Suggestion: we should introduce a path helper called `msvc_import_lib_for_dylib_name` or something like that

**Review by jieyouxu — APPROVED** [review]
Thanks, just a few suggestions

File: src/tools/run-make-support/src/artifact_names.rs [hunk 4]
```diff
@@ -38,6 +38,15 @@ pub fn dynamic_lib_name(name: &str) -> String {
     format!("{}{name}.{}", std::env::consts::DLL_PREFIX, std::env::consts::DLL_EXTENSION)
 }
 
+/// Construct the import dynamic library name, exclusive to MSVC and accepted by link.exe.
```
Comment (jieyouxu) [inline src/tools/run-make-support/src/artifact_names.rs:41]: Suggestion: "name of the import library for the dynamic library"

File: src/tools/run-make-support/src/artifact_names.rs [hunk 5]
```diff
@@ -38,6 +38,15 @@ pub fn dynamic_lib_name(name: &str) -> String {
...
+#[must_use]
+pub fn msvc_import_dynamic_lib_name(name: &str) -> String {
```
Comment (jieyouxu) [inline src/tools/run-make-support/src/artifact_names.rs:43]: Suggestion: `#[track_caller]`

File: src/tools/run-make-support/src/artifact_names.rs [hunk 6]
```diff
@@ -38,6 +38,15 @@ pub fn dynamic_lib_name(name: &str) -> String {
...
+    assert!(is_msvc(), "this function is exclusive to MSVC");
+    assert!(!name.contains(char::is_whitespace), "dynamic library name cannot contain whitespace");
```
Comment (jieyouxu) [inline src/tools/run-make-support/src/artifact_names.rs:45]: Nit: import library

**Review by jieyouxu — APPROVED** [review]
Thanks, a few more nits then r=me.

**Pull request merged**